core/src/main/scala/org/apache/spark/sql/execution/CodegenSparkFallback.scala

```diff
@@ -75,31 +100,40 @@ case class CodegenSparkFallback(var child: SparkPlan) extends UnaryExecNode {
         // is still usable:
         SystemFailure.checkFailure()
 
-        if (!isCodeGenerationException(t)) throw t
-
-        // fallback to Spark plan
-        val session = sqlContext.sparkSession.asInstanceOf[SnappySession]
-        session.getContextObject[() => QueryExecution](SnappySession.ExecutionKey) match {
-          case Some(exec) =>
-            val msg = new StringBuilder
-            var cause = t
-            while (cause ne null) {
-              if (msg.nonEmpty) msg.append(" => ")
-              msg.append(cause)
-              cause = cause.getCause
-            }
-            logInfo(s"SnappyData code generation failed due to $msg. Falling back to Spark plans.")
-            session.sessionState.disableStoreOptimizations = true
-            try {
-              val plan = exec().executedPlan
-              val result = f(plan)
-              // update child for future executions
-              child = plan
-              result
-            } finally {
-              session.sessionState.disableStoreOptimizations = false
-            }
-          case None => throw t
+        if (isConnectorCatalogStaleException(t)) {
+          logWarning(s"SmartConnector catalog is not upto date")
+          val session = sqlContext.sparkSession.asInstanceOf[SnappySession]
+          session.sessionCatalog.invalidateAll()
+          SnappySession.clearAllCache()
+          throw t
```

Comment:
I think we should throw a specialized exception, like CatalogStaleException, and ask the user to reconstruct the dataframe. 
Also we should try to make it transparent by recreating the CachedDataFrame but I understand the challenges. Please file a task and we will see later how to do this.
We may want to mention in docs about how user can handle the exception and instructions for retry.